feat: experimental raw env values flag (#12023)
withastro/docs: src/content/docs/en/reference/experimental-flags/raw-env-values.mdx

---
title: Experimental raw environment variables values
sidebar:

>   label: Raw environment variables values

Reviewer: ```suggestion
  label: Raw environment variables
```
For the label, this might be a little long? Would this still be meaningful enough?

i18nReady: true
---

import Since from '~/components/Since.astro'

<p>

**Type:** `boolean`<br />
**Default:** `false`<br />
<Since v="5.12.0" />
</p>

> The `experimental.rawEnvValues` flag disables coercion of `import.meta.env` values that are populated from `process.env`.

Reviewer: ```suggestion
Astro allows you to configure a [type-safe schema for your environment variables](/en/guides/environment-variables/#type-safe-environment-variables), and converts variables imported via `astro:env` into the expected type.  

However, Astro also additionally converts your environment variables imported from `import.meta.env`, and this can prevent access to some values such as the strings `"true"` (which is converted to a boolean value), and `"1"` (which is converted to a number).

The `experimental.rawEnvValues` flag disables coercion of `import.meta.env` values that are populated from `process.env`, allowing you to use the raw, imported value.
```
Once again making things up, but can we have an introduction maybe something like this?

> To disable coercion, set the flag to `true` in your Astro configuration:

Reviewer: ```suggestion
To disable Astro's default coercion on values imported from `import.meta.env`, set the `experimental.rawEnvValues` flag to `true` in your Astro configuration:
```

```js title="astro.config.mjs" ins={4-6}
import { defineConfig } from "astro/config"

export default defineConfig({
  experimental: {
    rawEnvValues: true,
  }
})
```

## Usage

> Currently, if an environment variable used through `import.meta.env` references a `process.env` variable whose value is `"true"`, `"false"`, `"1"`, or `"0"`, it will be coerced to respectively `true`, `false`, `1` and `0`. For example:
> 
> - If `import.meta.env.FOO` is called and the project root contains a `.env` file with `FOO=false`, the value will be `"false"` (string)
> - If `import.meta.env.FOO` is called and the dev server is started with `FOO=false astro dev`, the value will be `false` (boolean)
> 
> This behavior is a holdover from when Astro supported dynamic prerender values in routes, which was [removed in Astro v5](/en/guides/upgrade-to/v5/#removed-support-for-dynamic-prerender-values-in-routes).
> 
> 
> This experimental flag will keep the string value instead of turning into a boolean or a number. This aligns `import.meta.env`'s behavior in Astro with [Vite](https://vite.dev/guide/env-and-mode.html#env-variables).
> 
> In a future major version, Astro will switch to not coercing `import.meta.env` values by default, but you can opt in to the future behavior early using the `experimental.rawEnvValues` flag.
> 
> If you need coercion in Astro, we recommend you use [`astro:env`](/en/guides/environment-variables/).

Reviewer: ``````suggestion
Enabling this experimental flag will no longer convert string values into booleans or numbers. This aligns `import.meta.env`'s behavior in Astro with [Vite](https://vite.dev/guide/env-and-mode.html#env-variables).

In a future major version, Astro will switch to not coercing `import.meta.env` values by default, but you can opt in to the future behavior early using the `experimental.rawEnvValues` flag and if necessary, [updating your project](#updating-your-project) accordingly.

### Updating your project

If you were relying on this coercion, you may need to update your project to ensure that your `process.env` variables are now defined exactly as you want them to be used:

``` title=".env" del={1,4} ins={2,5}
REAL_EXAMPLE_PLEASE="false"
REAL_EXAMPLE_PLEASE=false

TIMEOUT="5"
TIMEOUT=5
```

Similarly, you may need to update your project code if you have created logic to work around Astro's default coercion:

```astro title="src/components/MyComponent.astro"
// SHOW CORRECTING SOME FAULTY LOGIC,
// LIKE PREVIOUSLY CHECKING FOR THE STRING TRUE OR FALSE 
// UPDATED TO NOW CHECK FOR THE BOOLEAN VALUE
```

If you need coercion in Astro, we recommend you use [`astro:env`](/en/guides/environment-variables/).
``````
How about something like this?

I don't think we need to go too far into a history lesson here. Rather, we can show what people might need to do to account for this change. This also helps us get a head start for when this becomes the default and we need to write the migration guide!

The `experimental.rawEnvValues` flag disables coercion of `import.meta.env` values that are populated from `process.env`, allowing you to use the raw value.

To disable Astro's coercion on values used through `import.meta.env`, set the `experimental.rawEnvValues` flag to `true` in your Astro configuration:

```js title="astro.config.mjs" ins={4-6}
import { defineConfig } from "astro/config"

export default defineConfig({
  experimental: {
    rawEnvValues: true,
  }
})
```

## Usage

Enabling this experimental flag will no longer convert string values into booleans or numbers. This aligns `import.meta.env`'s behavior in Astro with [Vite](https://vite.dev/guide/env-and-mode.html#env-variables).

In a future major version, Astro will switch to not coercing `import.meta.env` values by default, but you can opt in to the future behavior early using the `experimental.rawEnvValues` flag and if necessary, [updating your project](#updating-your-project) accordingly.

### Updating your project

> If you were relying on this coercion, you may need to update your project code to apply it manually:

Reviewer: This example is great! Just checking though, you would *not* expect people to be updating any environment variables themselves? Are these things they would typically not have access to, so the real workaround is in fixing the logic to use the raw value?

Author: Yeah they have no control over the env var they receive, it always be a string. So the only way really is to update were it's handled in code
